torchao/prototype/intx/intx.py

```diff
@@ -0,0 +1,630 @@
+import functools
+from collections import defaultdict
+from typing import Any, Callable, Dict, Optional, Tuple, Union, List
+
+import torch
+
+from torch.utils._python_dispatch import return_and_correct_aliasing
+from torchao.quantization.quant_primitives import choose_qparams_affine, MappingType
+from torchao.prototype.intx.bitpacking import pack, unpack, numbits
+from torchao.dtypes.utils import (
+    _implements,
+    _ATEN_OP_OR_TORCH_FN_TABLE,
+    _register_layout_cls,
+    _get_layout_tensor_constructor,
+)
+from torchao.quantization.quant_primitives import (
+    choose_qparams_affine,
+    quantize_affine,
+    dequantize_affine,
+    ZeroPointDomain,
+    MappingType,
+    int_scaled_matmul,
+)
+import pdb
+aten = torch.ops.aten
+
+class IntxLayout(torch.Tensor):
+    """
+    Base class for the layout tensor for `IntxTensor`
+    """
+    # this should be set for each layout class during registration
+    extended_layout: Optional[str] = None
+
+    def get_plain() -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        pass
+
+    @classmethod
+    def from_plain(
+        cls,
+        int_data: torch.Tensor,
+        scale: torch.Tensor,
+        zero_point: torch.Tensor,
+        bit_size: int,
+    ):
+        pass
+
+    def _get_to_kwargs(self, *args, **kwargs):
+        device, dtype, _, memory_format = torch._C._nn._parse_to(*args, **kwargs)
+        device = self.device if device is None else device
+        dtype = self.dtype if dtype is None else dtype
+        memory_format = (
+            memory_format if memory_format is not None else torch.preserve_format
+        )
+        kwargs = {
+            "device": device,
+            "dtype": dtype,
+            "memory_format": memory_format,
+        }
+        return kwargs
+    
+
+class IntxTensor(torch.Tensor):
+    """
+    Quantized tensor subclass for sub byte quantization (1-7 bits) with affine assymetric quantization
```

Comment:
this seems to be mixing the intx logic and affine quantization logic? could you split these?

intx_tensor (e.g. uint3) should be able to replace plain dtypes like torch.uint8 and compose with the rest of the affine quantization logic I think

Reply:
Yea I think I will separate out the quantization logic into the apply_tensor_subclass function. This should make the constructor a lot cleaner. I will also add some simple type checking in the Intx_tensor_class